### DISCUSSION TITLE
Dynamically Change AuthorizationUrl

### DISCUSSION THREAD

**Discussion by pardhav** [nextauthjs/next-auth#1998] [Help]
I'm trying to implement Salesforce Oauth. It consists of different Authorization URLs based on the environment we are trying to log in to. 

For Prod, it would be: https://login.salesforce.com/services/oauth2/token
For Sandbox, it would be:https://test.salesforce.com/services/oauth2/token

I tried creating a new custom provider, but the redirect_uri gets changed based on the unique id, hence salesforce throws validation error as redirect_uri mismatch

is there any way to change these URLs dynamically?

**Comment by balazsorban44** [comment]
have you tried using an environment variable?

BTW, you won't need to create a whole new custom provider, we have a built-in one, just override the values you want to https://next-auth.js.org/providers/salesforce#options

**Reply by pardhav** [reply]
i have been using built-in provider's itself. 

user story: 
user can either login to sandbox or production by selecting from a Drop-down. Based on the selection they must be logged in to respective environments

issue: 
i need to change the urls based on user selection, not on which environment my application is running hence environment variables doesn't solve my problem.

**Reply by balazsorban44** [reply]
What about
```js
providers: [
  Providers.Salesforce({
    clientId: process.env.SALESFORCE_ID,
    clientSecret: process.env.SALESFORCE_SECRET,
  }),
  Providers.Salesforce({
    id: "salesforce-sandbox",
    clientId: process.env.SALESFORCE_SANDBOX_ID,
    clientSecret: process.env.SALESFORCE_SANDBOX_SECRET,
    authorizationUrl: "https://test.salesforce.com/services/oauth2/token",
  }),
]
```

Then on your frontend you can do one of `signIn("salesforce")` or `signIn("salesforce-sandbox")`

**Reply by pardhav** [reply]
Thanks for the tip, but I was trying to avoid creating two different client id's just for logging into a different URL. The reason being I need to maintain 2 different client id's for each of my environments dev, staging, prod. I know that would be a one time setup but isn't there any other efficient way this can be achecived.

**Reply by balazsorban44** [reply]
You are essentially trying to talk to two distinct providers living on two different urls, even though they are the same type (Salesforce). I don't see another way than setting up two clients. you could have the same ID and secret for them, ID is probably fine but I don't recommend using the same secret for different providers. It should be unique for security, but ultimately I guess it would work with the same values.
